### ISSUE TITLE
pulumi {preview/up/...} is slow with medium sized TS stack

### ISSUE THREAD

**Issue by 0x53A** [pulumi/pulumi#7682]
I have a medium sized stack (181 total resources) with an AKS cluster, a few supporting azure resources, nginx-ingress, cert-manager, and a few other kubernetes resources.

-----------

I took a trace of a no-op up:

``pulumi up --skip-preview --tracing=file:./up.trace``

If you want to take a look, I can send you the details (update url and trace file [170MB]) per email or chat.
(note to self: it's update # 27)

A few things I noticed myself: there are *a lot* of calls to ``api/decryptValue``, each about 150ms to 200ms.

The ``pulumi`` trace alone is 1m 9s:

![image](https://user-images.githubusercontent.com/4236651/127551058-2ac1252e-78ac-4893-a2ff-286780996468.png)

Also, apparently the time logging is wrong/incomplete. In the console output it says ``Duration: 35s`` (I didn't stop the time myself), but in the trace file its over a minute.

Inside there, it's mostly ``/pulumirpc.ResourceMonitor/RegisterResource`` and ``/pulumirpc.ResourceMonitor/RegisterResource`` with 1s to 6s each.

![image](https://user-images.githubusercontent.com/4236651/127551669-875ee90d-4b75-434d-a624-fc0286838bd4.png)

--------

I'm on windows 10 x64 with pulumi 3.9.0.
The environment is Typescript / node.
I'm mostly using "@pulumi/azure-native" and "@pulumi/kubernetes".

-------

Edit: I ran ``pulumi up --skip-preview`` (no-op) again and stopped the time with my phone. It took 1m 4s in reality, the cli output said 32s.

----

Edit2: ``pulumi refresh --target i_do_not_exist`` takes about 30 seconds:

![image](https://user-images.githubusercontent.com/4236651/127558693-010a0564-12a1-42cc-b5db-ec41c71376b3.png)

**Comment by t0yv0** [comment]
Thanks for this report! We recently looked into TSC slowness in presence of '*' imports from azure-native (which is currently very large). Do you do top-level imports from azure-native? Targeted imports alleviate the situation a bit.

Also, the fact that console reports smaller timing than the trace is really troubling, I'd like to fix that.

**Comment by 0x53A** [comment]
I was doing

```ts
import { Provider, resources, storage, containerservice, network } from "@pulumi/azure-native";
```

``pulumi up --skip-preview`` took 1m 35s (and printed 1m 4s).

----------


and switched to

```ts
import * as resources from "@pulumi/azure-native/resources";
import * as storage from "@pulumi/azure-native/storage";
import * as containerservice from "@pulumi/azure-native/containerservice";
import * as network from "@pulumi/azure-native/network";
import { Provider } from "@pulumi/azure-native/provider";
```

``pulumi up --skip-preview`` took 1m 25s (and printed 52s).

--------

It's a little better, but might just be noise.

-------

also, I went ahead and sent the trace to your contact formular as request # 849 (since it's currently midnight over here and I had planned to go sleep sooner), FYI

Cameron pointed me to https://github.com/pulumi/pulumi/issues/6445, and I do have 97 calls to api/decryptValue.
Not sure how to count the total number of secrets though, I'm not creating any explicitly, just using Kubernetes.

**Comment by t0yv0** [comment]
Right, something else is going on here. I think the issue I thought of would affect you if you did `import *` from `@pulumi/azure-native`. Thanks for this report, we'll have to take a closer look.

**Comment by clstokes** [comment]
@0x53A can you share how long `pulumi stack` takes?

**Comment by 0x53A** [comment]
@clstokes 

> @0x53A can you share how long pulumi stack takes?

about 20 seconds

-----------

> cat ./stack.json | grep -c '4dabf18193072939515e22adb298388d'

returned ``98``.

**Comment by 0x53A** [comment]
So the DNS thing did help a bit, but after adding back nginx, cert-manager, etc, I'm still at 49s for a no-op ``pulumi up --skip-preview`` and at 1m 25s for a no-op ``pulumi up --yes``.

And this is with a mostly empty cluster, I expect there to be hundreds (or low thousands) of kubernetes objects (deployments, services, ingresses, ...).

Is this to be expected? Do you have any tips where I could look next?

**Comment by 0x53A** [comment]
The current state of my stack is:

1083 resources, 207 secrets.

``(Measure-Command { pulumi stack | Out-Default })``: 43 seconds
``(Measure-Command { pulumi preview | Out-Default })``: 15 minutes 52 seconds (probably a fluke)
``(Measure-Command { pulumi preview | Out-Default })`` 2 minutes 3 seconds
``pulumi up``: 2 minutes 10 seconds for the preview phase, 4 minutes 50 seconds total
``(Measure-Command { pulumi up --skip-preview --yes | Out-Default })``: 2 minutes 31 seconds

My internet connection is great, but I'm in Frankfurt, not in the US, so I have a lot of latency to the pulumi servers.

It always takes about 2 minutes until it starts to do anything, which I'd assume is the time it takes to download the current stack and decrypt the secrets. After this initial pause, performance is ok.

---------

about two minutes for a ``pulumi up --skip-preview`` is, well, not great, but still tolerable. The stack isn't expected to grow too much in the near future, but I have no idea about the distant future.


So I'm eagerly awaiting https://github.com/pulumi/pulumi/pull/8138 like a kid waiting for christmas.

Is there anything I can do for you, provide any logfiles, performance traces, etc?

**Comment by Kamaradeivanov** [comment]
I'm facing the same problems of slowness. I'm in France with strong professional fibre connection, but with a stack, used to manage my own gitlab server, with a little more than 1300 resources it's a nightmare.

For a simple `pulumi stack` it took more than 1 min
`pulumi stack  1.02s user 0.27s system 1% cpu 1:25.93 total`


For a preview, almost 4 min ...
`pulumi preview  50.00s user 4.99s system 23% cpu 3:55.35 total`

And recently I can’t apply any more, I encounter the following issue each time i try to do a `pulumi up`. 

```
error: resource complete event returned an error: failed to save snapshot: [403] The provided update token has expired.
error: update failed
```

**Comment by Kamaradeivanov** [comment]
@0x53A Have you find a solution ? I'm still stuck on a stack with 1300 resources... And no solution other than remove resources in my state file.... I'm still searching another way to manage secrets. I will probably use vault (hashicorp).

**Comment by 0x53A** [comment]
@Kamaradeivanov for quick iterative development I've created a VM in the US, though that is still relatively slow.

For "normal" prod updates, a few minutes is acceptable, we run it through a CI pipeline.


I haven't seen your error 403, can you maybe try to do a ``pulumi login`` again? Or create a support ticket, that's definitely a different issue.

**Comment by cveld** [comment]
@0x53A TF Provider -- you refer to the terraform bridge provider? Actually for local development I would expect webpack watch like performance. Press ctrl-s and get an instantaneous update in a sidebar. Why wait even 2 seconds. I am curious whether the pulumi core team has plans towards a webpack watch like architecture.

**Comment by twoTimesAgnew** [comment]
Has there been any movement on this? We've got multiple 250+ resource stacks and pulumi up takes ~2-12 mins on a TS stack...

**Comment by didlawowo** [comment]
have the same problem on mac M1 take a few minute only for 400 resources preview / up
